metadata functional improvements

mroz22 opened the pull request:
As szymon [pointed out](https://github.com/trezor/trezor-suite/pull/3027), dfd.reject is being called even on sucessful authorization flow. It actually did work but to make it more future proof I updated error handling in metadata provider modal.

should probably resolve #3207

szymonlesisz commented on packages/suite/src/services/suite/metadata/GoogleProvider.ts:
```diff
@@ -16,7 +16,7 @@ class GoogleProvider extends AbstractMetadataProvider {
             return true;
         } catch (error) {
-            return false;
+            return error.message as string;
```
```suggestion
            return error.toString();
```
what if message is not a string or there is no message?

mroz22 replied:
resolved with instanceof Error check

szymonlesisz commented on packages/suite/src/components/suite/modals/metadata/MetadataProvider/index.tsx:
```diff
@@ -57,11 +57,18 @@ const MetadataProvider = (props: Props) => {
+            // stop here, user might have changed his decision and wants to use another provider
             return;
         }
+        // TODO: we could show error in the UI
```
i would rather implement this "display error", at least we got some feedback if it happens and we know where to look.
its always better to receive a issue with "it's showing Error: some error" than "it isn't working, nothing happens" :)

mroz22 replied:
oook, I just don't know how it should look like. I may try something and if design people don't like it, they will come with something

mroz22 replied:
Indeed it looks like it does not break anything and lets me remove this eslint ignore, but lets wait for tests..

mroz22 replied:
Maybe something like this will do for now
![image](https://user-images.githubusercontent.com/30367552/100106472-ddf17f00-2e68-11eb-9ff4-aaa411666a7e.png)

mroz22 replied:
But I am wondering whether this text should be displayed to user at all :face_with_head_bandage: It clerly seems to be addressing me rather than user :no_good:

mroz22 commented:
Ok, while addressing review comments I found few more issues worth refactoring. So I did but I need to read my work tomorrow when I am fresh

mroz22 commented:
Reworked it a bit: 
- fixed problem with not unregistering event listeners 
- unified error handling from abstractProvider.connect method
- and probably more

tomexx commented on packages/suite/src/utils/suite/oauth.ts:
```diff
@@ -38,70 +38,125 @@ const openWindowOnAnotherDomain = (
-        const { code, access_token, state } = message;
+const getDesktopHandlerInstance = (
+    dfd: Deferred<Credentials>,
+    originalParams: { [key: string]: string },
```
let's maybe add type as it is used in multiple places

mroz22 replied:
well, why not. It has 13 occurrences in total across the project. What would be the name for this utility type?

keraf replied:
`KeyValueString`?

tomexx replied:
cache invalidation and naming stuff are the hardest things in software development 😅  anyway I'd go with @keraf 's suggestion

mroz22 replied:
And you would expect me to put it into types/utils/index.ts? I am still struggling to see added value of this. If I see `{[key:string]:string}` I immediately know what is going on, I see this one as a common idiom, nearly the same as `number` or `boolean`. When I see `KeyValueString` I really have no idea what to expect. It does not appear to be a good deal to have less typing 14 (KeyValueString) vs 21 ({[key:string]:string}) in exchange of added mental overhead. I'd support your idea in case it was more complicated type, where we could say that it might be edited in future. Here I don't see any chance how it could ever be edited.

mroz22 commented on packages/suite-desktop/scripts/dev.js:
```diff
@@ -14,7 +14,8 @@ next.prepare().then(() => {
     const server = createServer(requestHandler).listen(8000, () => {
         if (!launchElectron) {
-            return;
+            console.error('process.env.LAUNCH_ELECTRON not set to true')
```
Not connected with anything in this PR. I was just tinkering with env vars and bumbed into case when launchElectron was false and I ended up waiting forever

szymonlesisz commented on packages/suite/src/utils/suite/oauth.ts:
```diff
@@ -40,72 +40,125 @@ const openWindowOnAnotherDomain = (
-            dfd.reject(new Error('Cancelled'));
-        }
-        window.removeEventListener('message', onMessageWeb);
+        return handleResponse(
```
unnecessary return

szymonlesisz commented on packages/suite/src/utils/suite/oauth.ts:
```diff
@@ -40,72 +40,125 @@ const openWindowOnAnotherDomain = (
+    originalParams: { [key: string]: string },
+) => {
+    desktopHandlerInstance = message => {
+        return handleResponse(
```
unnecessary return

szymonlesisz commented on packages/suite-desktop/src-electron/preload.ts:
```diff
@@ -51,12 +51,17 @@ contextBridge.exposeInMainWorld('desktopApi', {
-
+    removeAllListeners: (channel: string) => {
+        ipcRenderer.removeAllListeners(channel);
+    },
```
implement this:
https://github.com/trezor/trezor-suite/pull/3175/files#diff-7a95c75c77571983a267b8413959e765e96268f9506d139a165c1253669818d7
